src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp

```diff
@@ -488,23 +488,50 @@ void mga::AtomicKMSOutput::wait_for_page_flip()
     pending_page_flip.get();
 }
 
-bool mga::AtomicKMSOutput::set_cursor(gbm_bo*)
-{
-    return false;
+bool mga::AtomicKMSOutput::set_cursor(gbm_bo* buffer)
+{
+    int result = 0;
+    if (configuration.lock()->current_crtc)
+    {
+        result = drmModeSetCursor(
+            drm_fd_,
+            configuration.lock()->current_crtc->crtc_id,
+            gbm_bo_get_handle(buffer).u32,
+            gbm_bo_get_width(buffer),
+            gbm_bo_get_height(buffer));
+        if (result)
+        {
+            mir::log_warning("set_cursor: drmModeSetCursor failed (%s)", strerror(-result));
+        }
+    }
+    return !result;
 }
 
-void mga::AtomicKMSOutput::move_cursor(geometry::Point)
+void mga::AtomicKMSOutput::move_cursor(geometry::Point destination)
 {
+    if (configuration.lock()->current_crtc)
+    {
+        if (auto result =
+                drmModeMoveCursor(drm_fd_, configuration.lock()->current_crtc->crtc_id, destination.x.as_int(), destination.y.as_int()))
+        {
+            mir::log_warning("move_cursor: drmModeMoveCursor failed (%s)", strerror(-result));
+        }
+    }
 }
 
 bool mga::AtomicKMSOutput::clear_cursor()
 {
-    return false;
-}
+    int result = 0;
+    if (configuration.lock()->current_crtc)
+    {
+        result = drmModeSetCursor(drm_fd_, configuration.lock()->current_crtc->crtc_id, 0, 0, 0);
```

Comment:
This is a TOCTOU¹ bug - because you're not *holding* the configuration lock, it's possible that another thread will null out `current_crtc` between your check and use.
```suggestion
    auto conf = configuration.lock()
    if (conf->current_crtc)
    {
        result = drmModeSetCursor(drm_fd_, conf->current_crtc->crtc_id, 0, 0, 0);
```

¹: Time Of Check to Time Of Use